### ISSUE TITLE
How do you tokenize the pose actions? 

### ISSUE THREAD

**Issue by amitkparekh** [vimalabs/VIMA#27]
I can see that you have the `discretize_actions` method in every "policy" class but you've not actually used it anywhere, nor have you used it within the example.py snippet. 

Can you please share a very specific code snippet that shows how you convert a single example from the raw dataset into a instance that is provided to the model during training. If you cannot use functions from the internal codebase, can you use pseudocode with strong typing so that we can see what each function/aspect is doing please?

**Comment by yunfanjiang** [comment]
Thank you for your interest in our project. To answer your questions:

> I can see that you have the discretize_actions method in every "policy" class but you've not actually used it anywhere, nor have you used it within the example.py snippet.

As mentioned in our paper, we directly encode actions in continuous format and output in discrete format. At training time, the `discretize_actions` method was used to provide targets/labels used in loss calculation. This codebase only provides inference-time behavior, where actions are already cached in discrete format. That's why this method is found unused.

> Can you please share a very specific code snippet that shows how you convert a single example from the raw dataset into an instance that is provided to the model during training.

Sure. Please first refer to functions [`prepare_prompt`](https://github.com/vimalabs/VIMA/blob/4758671789ce2b8e94181f917bc3dd948e866fe7/scripts/example.py#L243C1-L243C13) and [`prepare_obs`](https://github.com/vimalabs/VIMA/blob/4758671789ce2b8e94181f917bc3dd948e866fe7/scripts/example.py#L374) for how a single sample is prepared. After that, the method `discretize_actions` is used to provide targets, which are further used in `torch.nn.functional.cross_entropy` to calculate loss. Masks are used to avoid loss contributed from padded timesteps.

**Comment by yunfanjiang** [comment]
I figured out this [answer](https://github.com/vimalabs/VIMA/issues/28#issuecomment-1704011701) would be helpful as well.
